qemu/tests/boot_with_max_queues.py

```diff
@@ -0,0 +1,240 @@
+import logging
+import re
+import time
+
+from virttest import error_context
+from virttest import utils_test
+from virttest import utils_net
+from virttest import utils_misc
+from virttest.utils_windows import virtio_win
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Boot guest with different vectors, then do netperf testing.
+
+    1) Boot up VM with max vectors.
+    2) Check guest msi & queues info
+    3) Start 10 scp file transfer tests
+
+    :param test: QEMU test object.
+    :param params: Dictionary with the tes t parameters.
+    :param env: Dictionary with test environment.
+    """
+
+    def get_MSI_queue_from_traceview_output(output):
+        """
+        Extract msi&queues infomation from traceview log file output
+
+        :param output: the content of traceview processed log infomation
+        """
+
+        def _convert_fun(item):
+            value_pair = list(reversed(item.split()))
+            if not value_pair[1].isdigit():
+                value_pair[0] = 'MSIs'
+                value_pair[1] = 0
+            else:
+                value_pair[1] = int(value_pair[1])
+            return value_pair
+
+        info_str = "Start checking dump content for msi & queues info"
+        error_context.context(info_str, logging.info)
+        search_exp = r"\d+ MSIs|\d+ queues"
+        lines = output.split('\n')
+        for line in lines:
+            # special case for vectors = 0
+            queue_when_no_msi = re.search(r'No MSIX, using (\d+) queue', line)
+            if queue_when_no_msi:
+                return (0, int(queue_when_no_msi.group(1)))
+
+            search_res = re.findall(search_exp, line)
+            if not search_res:
+                continue
+            try:
+                value_dict = dict(map(_convert_fun, search_res))
+                MSIs_number = value_dict["MSIs"]
+                queues_number = value_dict["queues"]
+                return (MSIs_number, queues_number)
+            except Exception:
+                pass
+        return (None, None)
+
+    def _remove_file(session, file_path):
+        """
+        Remove a file at file_path
+
+        :param file_path: file path to remove
+        :param session: a logined session
+        """
+
+        clean_cmd = "del %s" % file_path
+        session.cmd_status(clean_cmd)
+
+    def _copy_file(session, src_path, dst_folder):
+        """
+        Copy a file at src_path to dst_folder.
+        If a same name file already exists at dst_folder, it will be replaced.
+
+        :param session: a logined sesison
+        :param src_path: file to copy from
+        :param dst_folder: folder to copy to
+
+        """
+
+        copy_cmd = "xcopy %s %s /y" % (src_path, dst_folder)
+        session.cmd_status(copy_cmd)
+
+    def get_MSIs_and_queues_windows(timeout=360):
```

Comment:
Do we have other test cases have similar steps? I mean, if we have multiple cases need same functions, better move them out of the `run` function, or even better we might include all such shared utils functions into one single module. Also for `prepare_traceview_windows`.